### ISSUE TITLE
After upgrading to 1.95.0, the "Source Control" panel no longer stays in sync with my git repo

### ISSUE THREAD

**Issue by kaiyoma** [microsoft/vscode#233129]
Type: <b>Bug</b>

Since upgrading, I've noticed that the "Source Control" panel does not stay in sync with my git repo. I find myself invoking "reload window" a lot throughout the day to fix the problem. This wasn't an issue before, but now I run into it constantly.

Example repro:
1. Add/modify files (through vscode or externally)
2. Source Control panel updates to show changes
3. Commit (from the command line)
4. Source Control panel still shows modified files until I "reload window"

VS Code version: Code 1.95.1 (65edc4939843c90c34d61f4ce11704f09d3e5cb6, 2024-10-31T05:14:54.222Z)
OS version: Windows_NT x64 10.0.22631
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz (8 x 2995)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled<br>webnn: disabled_off|
|Load (avg)|undefined|
|Memory (System)|15.71GB (2.63GB free)|
|Process Argv|--log trace --crash-reporter-id 4c14e0ab-6323-4be2-8ea1-b8ae8d1b35aa|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (10)</summary>

Extension|Author (truncated)|Version
---|---|---
copy-relative-path|ale|0.0.2
turbo-console-log|Cha|2.10.5
githistory|don|0.6.20
prettier-vscode|esb|11.0.0
gc-excelviewer|Gra|4.2.62
jenkins-declarative-support|jmM|0.1.0
code-groovy|mar|0.1.2
better-folding|Moh|0.5.1
color-highlight|nau|2.8.0
material-icon-theme|PKi|5.12.0

(1 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vspor879:30202332
vspor708:30202333
vspor363:30204092
vscod805:30301674
binariesv615:30325510
vsaa593cf:30376535
py29gd2263:31024239
c4g48928:30535728
azure-dev_surveyone:30548225
vscrpc:30673769
962ge761:30959799
pythongtdpath:30769146
pythonnoceb:30805159
asynctok:30898717
pythonmypyd1:30879173
h48ei257:31000450
pythontbext0:30879054
cppperfnew:31000557
dsvsc020:30976470
pythonait:31006305
dsvsc021:30996838
jg8ic977:31013176
dvdeprecation:31068756
dwnewjupytercf:31046870
impr_priority:31102340
nativerepl2:31139839
refactort:31108082
pythonrstrctxt:31112756
cf971741:31144450
iacca1:31171482
notype1cf:31157160
5fd0e150:31155592
dwcopilot:31170013

```

</details>

**Comment by kaiyoma** [comment]
Correction: after downgrading to 1.94.2, I am still seeing this, so it's not a new issue in 1.95.x.

This may be related to the fact that we use the "rebase" approach (we use Gerrit, not GitHub, for source control). When I notice this problem happening in vscode, I can see "Rebasing" in the lower left of the screen, even when I'm not in the middle of a rebase. After reloading the window, "Rebasing" (correctly) disappears and I see "trunk" instead.

**Comment by kaiyoma** [comment]
![Image](https://github.com/user-attachments/assets/58234f06-143a-4acd-bca5-3a6bf14a2338)

**Comment by kaiyoma** [comment]
In the screenshot above, you can see that `git status` is clean on the command line, but vscode is still showing that I have a lot of file edits. Sometimes this resolves itself, but usually I have to reload the window to get vscode to show the correct repo status.

It might be helpful to know that we follow the "Gerrit workflow" (not the GitHub workflow), whereby we are constantly doing interactive rebases, editing past commits, and then pushing those changes (as opposed to pushing new commits onto a fix branch).

**Comment by kaiyoma** [comment]
Here's another instance of vscode falling out of sync, but in the opposite direction:

![Image](https://github.com/user-attachments/assets/e8a777e5-dfb3-4d1e-8308-b40a99a64497)

The status bar says I'm rebasing, but I'm not (I already finished). In this case, I have staged git changes that are picked up by the command-line tools, but not being picked up by vscode. Again, reloading the window fixes things and vscode lists the files with changes.

**Comment by kaiyoma** [comment]
Updated title to reflect the fact that this happens with versions before 1.95.0 as well.

**Comment by Akinwunmi** [comment]
Facing the same issue after upgrading to 1.95.3.
Changes are not automatically reflected in Source Control, but only after clicking "Refresh".

**Comment by lszomoru** [comment]
The Source Control view is being refreshed by parsing the output if `git status` which is being executed as a response to file system events. Some git operations may result in a large number of file system events being emitted and in such cases it may take some time for VS Code to "catch up" but it should always do so. When things get out of sync could you using the "Refresh" action in the "Source Control" view? It would be interesting to see if that resolves the issue.

It would also be very helpful if you would find a scenario in which the issue can be consistently reproduced so that I can try to reproduce the issue and debug further.

**Comment by Akinwunmi** [comment]
In my case clearing the cache, rebooting my system, and disabling/enabling several extensions was the fix eventually.
I could recommend doing the same, hopefully one of the above or a combination will do the trick!

**Comment by kaiyoma** [comment]
> Some git operations may result in a large number of file system events being emitted and in such cases it may take some time for VS Code to "catch up" but it should always do so.

This seems like the right explanation for what I'm seeing, since I often make edits to dozens or hundreds of files and then rebuild possibly dozens of packages.

> When things get out of sync could you using the "Refresh" action in the "Source Control" view? It would be interesting to see if that resolves the issue.

Sure thing, I'll give this a shot the next time.

**Comment by lszomoru** [comment]
> I often make edits to dozens or hundreds of files and then rebuild possibly dozens of packages

Are the packages something that you track in source control?
If they are not you might want to look at the `files.watcherExclude` to exclude these from the file watcher.

**Comment by kaiyoma** [comment]
> Are the packages something that you track in source control? If they are not you might want to look at the `files.watcherExclude` to exclude these from the file watcher.

Yes, everything is in source control. `node_modules` is already being excluded from vscode, and that's really the only part of the directory tree that should be ignored.

**Comment by kaiyoma** [comment]
> > When things get out of sync could you using the "Refresh" action in the "Source Control" view? It would be interesting to see if that resolves the issue.
> 
> Sure thing, I'll give this a shot the next time.

Clicking the Refresh button fixes the problem immediately. It's not clear why it's not syncing on its own.

**Comment by lszomoru** [comment]
Thanks for confirming that pressing the "Refresh" button correctly updates the state of the "Source Control" view state. That only thing that I can think of is that when there is a large number of file-system events, they stop being delivered to the `vscode.git` extension hence we do not refresh the state of the view. 

To further investigate it would be great if we could come up with a "repository state" + a set of steps that we can use to reproduce the issue. If we have that, we can enable verbose logging and try to understand what happens with the file-system events.

**Comment by kaiyoma** [comment]
> That only thing that I can think of is that when there is a large number of file-system events, they stop being delivered to the `vscode.git` extension hence we do not refresh the state of the view.

I was able to repro this today without doing much. After finishing an interactive rebase where I made simple edits to a previous commit, I started editing a package by making a few new files and copying a few others. It was really light file system work, and the panel did not update until I clicked refresh. At this point, my best guess is that the interactive rebase process confuses vscode.

**Comment by lszomoru** [comment]
@kaiyoma, are you able to reproduce this in an isolated repository that you could share so that I can debug this? Thanks!

**Comment by kaiyoma** [comment]
> [@kaiyoma](https://github.com/kaiyoma), are you able to reproduce this in an isolated repository that you could share so that I can debug this? Thanks!

No, probably not. It seems to happen randomly as I work in our repo.
